UserTools/MeanTimeCheck/MeanTimeCheck.cpp

```diff
@@ -0,0 +1,185 @@
+	double WeightedAverage = this->GetWeightedAverage();
+	double WeightedPeak = this-> GetWeightedPeak();
+std::cout<<"Check 2"<<std::endl;
+
+	peakTime->Fill(PeakTime);
+	weightedAverage->Fill(WeightedAverage);
+	weightedPeak->Fill(WeightedPeak);
+std::cout<<"Check 3"<<std::endl;
+	m_data->Stores.at("RecoEvent")->Set("meanTime", PeakTime);
+
+	return true;
+
+}
+
+
+bool MeanTimeCheck::Finalise(){
+
+	Output_tfile->cd();
+	Output_tfile->Write();
+	Output_tfile->Close();
+	Log("MeanTimeCheck exitting", v_debug, verbosity);
+
+  return true;
+}
+
+double MeanTimeCheck::GetPeakTime() {
+	return delta->GetBinCenter(delta->GetMaximumBin());
+}
+
+double MeanTimeCheck::GetWeightedAverage() {
+	double Swx = 0.0;
+	double Sw = 0.0;
+
+	double delta = 0.0;
+	double sigma = 0.0;
+	double weight = 0.0;
+	double deweight = 0.0;
+	double deltaAngle = 0.0;
+	double meanTime = 0.0;
+
+	double myConeEdgeSigma = 7.0;  // [degrees]
+
+	for (int idigit = 0; idigit < this->VtxGeo->GetNDigits(); idigit++) {
+		int detType = this->VtxGeo->GetDigitType(idigit);
+		if (this->VtxGeo->IsFiltered(idigit)) {
+			delta = this->VtxGeo->GetDelta(idigit);
+			sigma = this->VtxGeo->GetDeltaSigma(idigit);
+			weight = 1.0 / (sigma*sigma);
+			// profile in angle
+			deltaAngle = this->VtxGeo->GetAngle(idigit) - Parameters::CherenkovAngle();
+			// deweight hits outside cone
+			if (deltaAngle <= 0.0) {
+			deweight = 1.0;
+			}
+			else {
+			deweight = 1.0 / (1.0 + (deltaAngle*deltaAngle) / (myConeEdgeSigma*myConeEdgeSigma));
+			}
+			Swx += deweight * weight*delta; //delta is expected vertex time 
+			Sw += deweight * weight;
+			}
+			}
+			if (Sw > 0.0) {
+			meanTime = Swx * 1.0 / Sw;
+			}
+			
+return meanTime;
+}
+
+double MeanTimeCheck::GetWeightedPeak() {
+	double sigma = 0.0;
+	double deltaAngle = 0.0;
+	double weight = 0.0;
+	double deweight = 0.0;
+	double myConeEdgeSigma = 7.0;  // [degrees]
+	double meanTime = 0.0;
+	vector<double> deltaTime1;
+	vector<double> deltaTime2;
+	vector<double> TimeWeight;
+
+	for (int idigit = 0; idigit < this->VtxGeo->GetNDigits(); idigit++) {
+		if (this->VtxGeo->IsFiltered(idigit)) {
+			deltaTime1.push_back(this->VtxGeo->GetDelta(idigit));
+			deltaTime2.push_back(this->VtxGeo->GetDelta(idigit));
+			sigma = this->VtxGeo->GetDeltaSigma(idigit);
+			weight = 1.0 / (sigma*sigma);
+			deltaAngle = this->VtxGeo->GetAngle(idigit) - Parameters::CherenkovAngle();
+			if (deltaAngle <= 0.0) {
+				deweight = 1.0;
+			}
+			else {
+				deweight = 1.0 / (1.0 + (deltaAngle*deltaAngle) / (myConeEdgeSigma*myConeEdgeSigma));
+			}
+			TimeWeight.push_back(deweight*weight);
+		}
+	}
+	int n = deltaTime1.size();
+	std::sort(deltaTime1.begin(), deltaTime1.end());
+	double timeMin = deltaTime1.at(int((n - 1)*0.05)); // 5% of the total entries
+	double timeMax = deltaTime1.at(int((n - 1)*0.90)); // 90% of the total entries
+	int nbins = int(n / 5);
+	TH1D *hDeltaTime = new TH1D("hDeltaTime", "hDeltaTime", nbins, timeMin, timeMax);
```

Comment:
why does this need to be on the heap? heap allocation is slow and risks memory leaks, this could just as easily be declared on the stack with no such issues.